Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -43,31 +58,48 @@ public struct BSONDocumentIterator: IteratorProtocol {
         }
 
         guard let type = BSONType(rawValue: typeByte), type != .invalid else {
-            fatalError("BSONDocumentIterator Failed: Invalid type, \(typeByte)")
+            throw BSONIterationError(
+                buffer: self.buffer,
+                typeByte: typeByte,
+                message: "Invalid type indicator"
+            )
         }
 
-        do {
-            let key = try self.buffer.readCString()
-            guard let bson = try BSON.allBSONTypes[type]?.read(from: &buffer) else {
-                throw BSONError.InternalError(message: "Cannot read unknown type: \(type)")
-            }
-            return (key: key, value: bson)
-        } catch {
-            fatalError("BSONDocumentIterator.next() failed: \(error)")
+        let key = try self.buffer.readCString()
+        guard let bson = try BSON.allBSONTypes[type]?.read(from: &buffer) else {
+            throw BSONIterationError(
+                buffer: self.buffer,
+                key: key,
+                type: type,
+                typeByte: typeByte,
+                message: "Cannot decode type"
+            )
         }
+        return (key: key, value: bson)
     }
 
     /// Finds the key in the underlying buffer, and returns the [startIndex, endIndex) containing the corresponding
     /// element.
-    internal func findByteRange(for key: String) -> (startIndex: Int, endIndex: Int) {
-        fatalError("Unimplemented")
+    internal mutating func findByteRange(for searchKey: String) -> (startIndex: Int, endIndex: Int)? {
```

Comment:
Instead of our own tuple type, we can use the stdlib [`Range<T>`](https://developer.apple.com/documentation/swift/range) and return a `Range<Int>?`.